File: forecasting-sdk-sample/forecasting-app.py
import time
from oci.ai_forecasting.forecast_client import ForecastClient
from oci.ai_forecasting.models import DataSourceDataAsset, FcDataSourceDataAsset, \
    ForecastCreationDetails, Forecast, CreateProjectDetails, DataSourceDetails, DataSourceDetailsObjectStorage, \
    CreateDataAssetDetails,UpdateForecastDetails
from oci.ai_forecasting.models.create_forecast_details import CreateForecastDetails
from oci.ai_forecasting.models.model_details import ModelDetails
from oci.ai_forecasting.models.schema import Schema
from oci.config import from_file

# change the config file location
config = from_file("/Users/rommonda/.oci/config")

compartment_id = "ocid1.compartment.oc1..aaaaaaaaa6b6bypmwsy5z3mhwdfdnc4mlt35xn5solpepqd6slplonok7lvq"
forecast_client = ForecastClient(
    config,
    service_endpoint="https://forecasting-int.aiservice.us-ashburn-1.oci.oraclecloud.com",
)

# CREATE PROJECT
print("-*-*-*-PROJECT-*-*-*-")

proj_details = CreateProjectDetails(
    display_name="PythonSDKTestProject",
    description="PythonSDKTestProject description",
    compartment_id=compartment_id,
)
create_res = forecast_client.create_project(create_project_details=proj_details)
print("----Creating Project----")
print(create_res.data)
time.sleep(5)
project_id = create_res.data.id

# GET PROJECT
get_proj = forecast_client.get_project(project_id=project_id)
print("----Get Project---")
print(get_proj.data)
time.sleep(5)

# CREATE DATA ASSET
print("-*-*-*-DATA ASSET-*-*-*-")
dDetails = DataSourceDetails(data_source_type="ORACLE_OBJECT_STORAGE")

dObjDetails = DataSourceDetailsObjectStorage(
    namespace="axebztqviz4z",
    bucket_name="testautomation",
    object_name="rossmann_10x500_primary.csv",
)

da_details = CreateDataAssetDetails(
    display_name="PythonSDKTestDataAsset",
    description="SDK Test DataAsset",
    compartment_id=compartment_id,
    project_id=project_id,
    data_source_details=dObjDetails,
)
create_res = forecast_client.create_data_asset(create_data_asset_details=da_details)
print("----Creating DataAsset----")
print(create_res.data)
time.sleep(5)
da_id = create_res.data.id

# GET DATAASSET
get_da = forecast_client.get_data_asset(data_asset_id=da_id)
print("----Reading DataAsset----")
print(get_da.data)
time.sleep(5)

# CREATE FORECAST
print("-*-*-*-Forecast-*-*-*-")

primaryFcDataSource = FcDataSourceDataAsset(
    data_asset_id=da_id,
    column_schema=[Schema(column_name="date", data_type=Schema.DATA_TYPE_DATE),
                   Schema(column_name="Sales", data_type=Schema.DATA_TYPE_INT),
                   Schema(column_name="Store", data_type=Schema.DATA_TYPE_STRING)],
    is_data_grouped=True,
    data_frequency="1DAY",
    ts_col_name="date",
    ts_col_format="yyyy-MM-dd HH:mm:ss"
)

dataSourceDataAsset = DataSourceDataAsset(
    primary_data_source=primaryFcDataSource
)

forecastCreationDetails = ForecastCreationDetails(
    data_source_details=dataSourceDataAsset,
    model_details=ModelDetails(models = ["SMA", "DMA"]),
    forecast_frequency="1WEEK",
    forecast_technique=ForecastCreationDetails.FORECAST_TECHNIQUE_ROCV,
    is_forecast_explanation_required=True,
    confidence_interval=95,
    target_variables = ["Sales"],
    forecast_horizon = 14,
    error_measure = ForecastCreationDetails.ERROR_MEASURE_RMSE,
    are_csv_result_files_required=True
)
createForecastDetails = CreateForecastDetails(
    display_name="PythonSDKTestForecast",
    description="PythonSDKTestForecast description",
    compartment_id=compartment_id,
    project_id=project_id,
    forecast_creation_details=forecastCreationDetails,
)

print("----Creating Forecast----")
create_res = forecast_client.create_forecast(create_forecast_details=createForecastDetails)
print(create_res.data)
time.sleep(5)
forecast_id = create_res.data.id

# GET FORECAST
get_forecast = forecast_client.get_forecast(forecast_id=forecast_id)
print("----Get Forecast----")
print(get_forecast.data)
time.sleep(5)
while get_forecast.data.lifecycle_state == Forecast.LIFECYCLE_STATE_CREATING:
    get_forecast = forecast_client.get_forecast(forecast_id=forecast_id)
    time.sleep(5)
    print("Status: "+ get_forecast.data.lifecycle_state)

# UPDATE FORECAST
print("----Updating Forecast----")

update_details = UpdateForecastDetails(
    display_name="Updated Forecast",
    description="Updated Forecast Description through SDK"
)

update_res = forecast_client.update_forecast(forecast_id=forecast_id,update_forecast_details=update_details)
print(update_res.data)

get_forecast = forecast_client.get_forecast(forecast_id=forecast_id)
time.sleep(1)
while get_forecast.data.lifecycle_state == Forecast.LIFECYCLE_STATE_UPDATING:
    get_forecast = forecast_client.get_forecast(forecast_id=forecast_id)
    time.sleep(1)
    print("Status: "+ get_forecast.data.lifecycle_state)

# GET RESULTS OF FORECAST
print("----Get STATUS of Forecast----")
get_forecast = forecast_client.get_forecast(forecast_id=forecast_id)
print(get_forecast.data)

print("----Get Forecast Result Content of Forecast----")
forecast_content = forecast_client.get_forecast_content(forecast_id=forecast_id,series_id=None)
print(forecast_content.data)

print("----Get Explanation Content of Forecast----")
explanation_content = forecast_client.get_explanation_content(forecast_id=forecast_id,series_id=None)
print(explanation_content.data)

print("----Get FittedSeries Content of Forecast----")
fittedSeries = forecast_client.get_fitted_series_content(forecast_id=forecast_id,series_id=None)
print(fittedSeries.data)

print("----Get InputSeries Content of Forecast----")
inputSeries = forecast_client.get_input_series(forecast_id=forecast_id,series_id=None)
print(inputSeries.data)

print("----Download Result CSV as ZIP----")
res = forecast_client.get_result_content(forecast_id=forecast_id)
file = open('response.zip', 'wb')
file.write(res.data.content)
file.close()

# DELETE FORECAST
delete_forecast = forecast_client.delete_forecast(forecast_id=forecast_id)
print("----Deleting Forecast----")
if delete_forecast.status == 204:
    print("----Forecast Deleted Successfully----")
else:
    print("----Forecast Deletion Failed----")
    print(delete_forecast.data)
time.sleep(2)

# CREATE FORECAST "ToBeCancelled"
createForecastDetails = CreateForecastDetails(
    display_name="PythonSDKTestForecast_ToBeCancelled",
    description="PythonSDKTestForecast_ToBeCancelled",
    compartment_id=compartment_id,
    project_id=project_id,
    forecast_creation_details=forecastCreationDetails,
)

create_res = forecast_client.create_forecast(create_forecast_details=createForecastDetails)
print("----Creating ToBeCancelled Forecast----")
print(create_res.data)
time.sleep(5)
cancel_forecast_id = create_res.data.id

#CANCEL FORECAST "ToBeCancelled"
cancel_forecast = forecast_client.cancel_forecast(forecast_id=cancel_forecast_id)
print("----Cancelling Forecast----")
if cancel_forecast.status == 202:
    get_forecast=forecast_client.get_forecast(forecast_id=cancel_forecast_id)
    while get_forecast.data.lifecycle_state == Forecast.LIFECYCLE_STATE_CANCELING:
        print('Canceling ...')
        time.sleep(2)
        get_forecast=forecast_client.get_forecast(forecast_id=cancel_forecast_id)
    if get_forecast.data.lifecycle_state == Forecast.LIFECYCLE_STATE_CANCELED:
        print("----Forecast Cancelled Successfully----")
    else:
        print("----Forecast Cancellation Failed----")
else:
    print("----Forecast Cancellation Failed----")
    print(cancel_forecast.data)

# DELETE FORECAST
delete_forecast = forecast_client.delete_forecast(forecast_id=cancel_forecast_id)
print("----Deleting ToBeCancelled Forecast----")
if delete_forecast.status == 204:
    print("----Forecast Deleted Successfully----")
else:
    print("----Forecast Deletion Failed----")
    print(delete_forecast.data)
time.sleep(2)

# DELETE DATA ASSET
delete_da = forecast_client.delete_data_asset(data_asset_id=da_id)
print("----DELETING DATA ASSET----")
if delete_da.status == 204:
    print("----DataAsset Deleted Successfully----")
else:
    print("----DataAsset Deletion Failed----")
    print(delete_da.data)
time.sleep(5)

# DELETE PROJECT
print("----DELETING PROJECT----")
delete_project = forecast_client.delete_project(project_id=project_id)
if delete_project.status == 204:
    print("----Project Deleted Successfully----")
else:
    print("----Project Deletion Failed----")
    print(delete_project.data)

File: forecasting-sdk-sample/forecasting-app.java
package org.ocasfc;

import com.oracle.bmc.ConfigFileReader;
import com.oracle.bmc.aiforecasting.Forecast;
import com.oracle.bmc.aiforecasting.ForecastClient;
//import com.oracle.bmc.aiforecasting.model.ConfidenceInterval;
import com.oracle.bmc.aiforecasting.model.*;
//import com.oracle.bmc.aiforecasting.model.ForecastFrequency;
import com.oracle.bmc.aiforecasting.requests.CreateDataAssetRequest;
import com.oracle.bmc.aiforecasting.requests.CreateForecastRequest;
import com.oracle.bmc.aiforecasting.requests.CreateProjectRequest;
import com.oracle.bmc.aiforecasting.requests.DeleteDataAssetRequest;
import com.oracle.bmc.aiforecasting.requests.UpdateForecastRequest;
import com.oracle.bmc.aiforecasting.requests.DeleteForecastRequest;
import com.oracle.bmc.aiforecasting.requests.CancelForecastRequest;
import com.oracle.bmc.aiforecasting.requests.DeleteProjectRequest;
import com.oracle.bmc.aiforecasting.requests.GetForecastRequest;
import com.oracle.bmc.aiforecasting.requests.GetForecastContentRequest;
import com.oracle.bmc.aiforecasting.requests.GetExplanationContentRequest;
import com.oracle.bmc.aiforecasting.requests.GetFittedSeriesContentRequest;
import com.oracle.bmc.aiforecasting.requests.GetInputSeriesRequest;
import com.oracle.bmc.aiforecasting.requests.GetResultContentRequest;
import com.oracle.bmc.aiforecasting.responses.CreateDataAssetResponse;
import com.oracle.bmc.aiforecasting.responses.CreateForecastResponse;
import com.oracle.bmc.aiforecasting.responses.CreateProjectResponse;
import com.oracle.bmc.aiforecasting.responses.GetForecastResponse;
import com.oracle.bmc.aiforecasting.responses.GetForecastContentResponse;
import com.oracle.bmc.aiforecasting.responses.GetExplanationContentResponse;
import com.oracle.bmc.aiforecasting.responses.GetFittedSeriesContentResponse;
import com.oracle.bmc.aiforecasting.responses.GetInputSeriesResponse;
import com.oracle.bmc.aiforecasting.responses.GetResultContentResponse;
import com.oracle.bmc.aiforecasting.responses.UpdateForecastResponse;
import com.oracle.bmc.aiforecasting.responses.DeleteForecastResponse;
import com.oracle.bmc.aiforecasting.responses.DeleteDataAssetResponse;
import com.oracle.bmc.aiforecasting.responses.DeleteProjectResponse;
import com.oracle.bmc.aiforecasting.responses.CancelForecastResponse;
import com.oracle.bmc.auth.AuthenticationDetailsProvider;
import com.oracle.bmc.auth.ConfigFileAuthenticationDetailsProvider;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.Arrays;

public class Application {

    private static final String compartmentId = "ocid1.tenancy.oc1..aaaaaaaaonxo6xgq5hmmbqmxqthumokfzgkdz6nje5g3mfoqq37azuuui3ea";

    public static void main(String[] args) throws IOException {

        final ConfigFileReader.ConfigFile configFile = ConfigFileReader.parse("/Users/rommonda/.oci/config");
        final AuthenticationDetailsProvider provider = new ConfigFileAuthenticationDetailsProvider(configFile);

        /* Create forecast service client */
        ForecastClient forecastClient = new ForecastClient(provider);
        forecastClient.setEndpoint("https://forecasting-int.aiservice.us-ashburn-1.oci.oraclecloud.com");

        //Create Project
        String projectId = createProject(forecastClient);
        //Create DataAsset
        String dataAssetId = createDataAsset(forecastClient, projectId);
        //Create Forecast
        String forecastId = createForecast(forecastClient, projectId, dataAssetId);
        try {
            Thread.sleep(10000);
            System.out.println("-----Get Forecast------");
            String forecastStatus = getForecast(forecastClient, forecastId);
            while (forecastStatus.equalsIgnoreCase("CREATING")) {
                Thread.sleep(2000);
                forecastStatus = getForecast(forecastClient, forecastId);
            }
            if (forecastStatus.equalsIgnoreCase("ACTIVE")) {
                System.out.println("-----Forecast Created Successfully------");
            } else {
                System.out.println("-----Forecast Creation Failed------");
            }
            Thread.sleep(10000);

            //Updating Forecast
            System.out.println("-----Updating Forecast------");
            UpdateForecastDetails updateForecastDetails = UpdateForecastDetails.builder().displayName("Updated Forecast").description("Updated Forecast Description through SDK").build();
            UpdateForecastResponse updateForecastResponse = forecastClient.updateForecast(UpdateForecastRequest.builder().forecastId(forecastId).updateForecastDetails(updateForecastDetails).build());
            System.out.println(updateForecastResponse.getForecast().toString());
            String updateForecastStatus = getForecast(forecastClient,forecastId);
            while ( updateForecastStatus.equalsIgnoreCase("UPDATING") ) {
                Thread.sleep(2000);
                updateForecastStatus = getForecast(forecastClient, forecastId);
            }

            //Get Results of Forecast
            System.out.println("-----Get Forecast Result Content of Forecast------");
            GetForecastContentResponse getForecastContentResponse = forecastClient.getForecastContent(GetForecastContentRequest.builder().forecastId(forecastId).seriesId("1063").build());
            System.out.println(getForecastContentResponse.getForecastContent().toString());

            System.out.println("-----Get Explanation Content of Forecast------");
            GetExplanationContentResponse getExplanationContentResponse = forecastClient.getExplanationContent(GetExplanationContentRequest.builder().forecastId(forecastId).seriesId("1063").build());
            System.out.println(getExplanationContentResponse.getExplanationContent().toString());

            System.out.println("-----Get FittedSeries Content of Forecast------");
            GetFittedSeriesContentResponse getFittedSeriesContentResponse = forecastClient.getFittedSeriesContent(GetFittedSeriesContentRequest.builder().forecastId(forecastId).seriesId("1063").build());
            System.out.println(getFittedSeriesContentResponse.getFittedSeriesContent().toString());

            System.out.println("-----Get InputSeries Content of Forecast------");
            GetInputSeriesResponse getInputSeriesResponse = forecastClient.getInputSeries(GetInputSeriesRequest.builder().forecastId(forecastId).seriesId("1063").build());
            System.out.println(getInputSeriesResponse.getInputSeries().toString());

            System.out.println("----Download Result CSV as ZIP----");
            GetResultContentResponse getResultContentResponse = forecastClient.getResultContent(GetResultContentRequest.builder().forecastId(forecastId).build());
            System.out.println(getResultContentResponse.get__httpStatusCode__());
            InputStream initialStream = getResultContentResponse.getInputStream();
            File targetFile = new File("response.zip");
            System.out.println("Downloading as ZIP...");
            Files.copy(initialStream, targetFile.toPath(), StandardCopyOption.REPLACE_EXISTING);


            //Deleting Forecast
            System.out.println("-----Deleting Forecast------");
            DeleteForecastResponse deleteForecastResponse=forecastClient.deleteForecast(DeleteForecastRequest.builder().forecastId(forecastId).build());
            if (deleteForecastResponse.get__httpStatusCode__() == 204)
                System.out.println("Forecast Deleted Successfully");
            else
                System.out.println("Forecast Deletion Failed");
            Thread.sleep(10000);

            //Create ToBeCancelled Forecast
            String cancelForecastId = createForecast(forecastClient, projectId, dataAssetId);

            //Cancel ToBeCancelled Forecast
            System.out.println("-----Canceling Forecast------");
            CancelForecastResponse cancelForecastResponse = forecastClient.cancelForecast(CancelForecastRequest.builder().forecastId(cancelForecastId).build());
            if (cancelForecastResponse.get__httpStatusCode__() == 202) {
                String cancelForecastStatus = getForecast(forecastClient, cancelForecastId);
                while (cancelForecastStatus.equalsIgnoreCase("CANCELING")) {
                    System.out.println("Canceling ...");
                    Thread.sleep(2000);
                    cancelForecastStatus = getForecast(forecastClient, cancelForecastId);
                }
                if (cancelForecastStatus.equalsIgnoreCase("CANCELED"))
                    System.out.println("Forecast Canceled Successfully");
                else
                    System.out.println("Forecast Cancellation Failed");
            }
            else {
                System.out.println("Forecast Cancellation Failed");
            }

            //Deleting ToBeCancelled Forecast
            System.out.println("-----Deleting Forecast------");
            DeleteForecastResponse deleteToBeCancelledForecastResponse = forecastClient.deleteForecast(DeleteForecastRequest.builder().forecastId(cancelForecastId).build());
            if (deleteToBeCancelledForecastResponse.get__httpStatusCode__() == 204)
                System.out.println("Forecast Deleted Successfully");
            else
                System.out.println("Forecast Deletion Failed");
            Thread.sleep(10000);

            //Deleting DataAsset
            System.out.println("-----Deleting DataAsset------");
            DeleteDataAssetResponse deleteDataAssetResponse = forecastClient.deleteDataAsset(DeleteDataAssetRequest.builder().dataAssetId(dataAssetId).build());
            if (deleteDataAssetResponse.get__httpStatusCode__() == 204)
                System.out.println("DataAsset Deleted Successfully");
            else
                System.out.println("DataAsset Deletion Failed");
            Thread.sleep(10000);

            //Deleting Project
            System.out.println("-----Deleting Project------");
            DeleteProjectResponse deleteProjectResponse = forecastClient.deleteProject(DeleteProjectRequest.builder().projectId(projectId).build());
            if (deleteProjectResponse.get__httpStatusCode__() == 204)
                System.out.println("Project Deleted Successfully");
            else
                System.out.println("Project Deletion Failed");
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public static String createProject(ForecastClient forecastClient) {
        CreateProjectDetails createProjectDetails = CreateProjectDetails.builder()
                .displayName("Testing Project From SDK " + System.currentTimeMillis())
                .description("Testing Forecasting Project from Java SDK")
                .compartmentId(compartmentId)
                .build();
        CreateProjectRequest createProjectRequest = CreateProjectRequest.builder()
                .createProjectDetails(createProjectDetails).build();
        CreateProjectResponse createProjectResponse = forecastClient.createProject(createProjectRequest);
        System.out.println("-----Creating Project------");
        System.out.println(createProjectResponse);
        System.out.println("-----Project Created------");
        return createProjectResponse.getProject().getId();
    }

    public static String createDataAsset(ForecastClient forecastClient, String projectId) {
        DataSourceDetailsObjectStorage dataSourceDetailsObjectStorage = DataSourceDetailsObjectStorage.builder()
                .objectName("rossmann_10x500_primary.csv")
                .bucketName("testautomation")
                .namespace("axebztqviz4z")
                .build();

        CreateDataAssetDetails createDataAssetDetails = CreateDataAssetDetails.builder()
                .dataSourceDetails(dataSourceDetailsObjectStorage)
                .compartmentId(compartmentId)
                .displayName("Cloud Capacity Data Asset By Java SDK")
                .projectId(projectId)
                .description("Data Asset for Java SDK testing")
                .build();

        CreateDataAssetRequest createDataAssetRequest = CreateDataAssetRequest.builder()
                .createDataAssetDetails(createDataAssetDetails).build();
        CreateDataAssetResponse createDataAssetResponse = forecastClient.createDataAsset(createDataAssetRequest);
        System.out.println("-----Creating DataAsset------");
        System.out.println(createDataAssetResponse);
        System.out.println("-----DataAsset Created------");
        return createDataAssetResponse.getDataAsset().getId();
    }

    public static String getForecast(ForecastClient forecastClient, String forecastId) {
        GetForecastRequest getForecastRequest = GetForecastRequest.builder()
                .forecastId(forecastId)
                .build();
        GetForecastResponse getForecastResponse = forecastClient.getForecast(getForecastRequest);
        System.out.println(getForecastResponse.getForecast());
        System.out.println("Status: "+getForecastResponse.getForecast().getLifecycleState().toString());
        return getForecastResponse.getForecast().getLifecycleState().toString();
    }

    public static String createForecast(ForecastClient forecastClient, String projectId, String dataAssetId) {
        System.out.println("-----Creating Forecast------");
        CreateForecastDetails createForecastDetails = CreateForecastDetails.builder()
                .forecastCreationDetails(getForecastCreationDetails(dataAssetId))
                .compartmentId(compartmentId)
                .projectId(projectId)
                .displayName("Forecast By Java SDK")
                .description("Forecast By Java SDK For Testing")
                .build();

        CreateForecastResponse createForecastResponse = forecastClient.createForecast(
                CreateForecastRequest.builder().createForecastDetails(createForecastDetails).build());
        System.out.println(createForecastResponse);
        System.out.println("-----Forecast Request Created------");
        return createForecastResponse.getForecast().getId();
    }

    private static ForecastCreationDetails getForecastCreationDetails(String dataAssetId) {
        return ForecastCreationDetails.builder()
                .dataSourceDetails(getDataSourceDetails(dataAssetId))
                .modelDetails(ModelDetails.builder().models(Arrays.asList(Models.Sma, Models.Dma)).build())
                .forecastFrequency("1DAY")
                .forecastHorizon(4)
                .forecastTechnique(ForecastTechnique.Rocv)
                .isForecastExplanationRequired(true)
                .errorMeasure(ErrorMeasure.Rmse)
                .targetVariables(Arrays.asList("Sales"))
                .confidenceInterval(95)
                .areCSVResultFilesRequired(true)
                .build();
    }

    private static ForecastDataSourceDetails getDataSourceDetails(String dataAssetId) {
        return DataSourceDataAsset.builder()
                .primaryDataSource(FcDataSourceDataAsset.builder()
                        .dataAssetId(dataAssetId)
                        .columnSchema(Arrays.asList(Schema.builder().columnName("date").dataType(DataTypes.Date).build(),
                                Schema.builder().columnName("Sales").dataType(DataTypes.Int).build(),
                                Schema.builder().columnName("Store").dataType(DataTypes.String).build()))
                        .isDataGrouped(true)
                        .dataFrequency("1DAY")
                        .tsColFormat("yyyy-MM-dd HH:mm:ss")
                        .tsColName("date")
                        .build())
                .build();
    }
}